lib/cn/kvset_split.c

```diff
@@ -482,87 +597,143 @@ vblocks_split(
+
+                err = mpool_mblock_props_get(ks->ks_mp, mbid, &props);
+                if (ev(err))
+                    goto out;
+
+                blks_right->bl_vused -= props.mpr_write_len - props.mpr_alloc_cap;
+                blks_left->bl_vtotal += props.mpr_alloc_cap - VBLOCK_FOOTER_LEN;
+            } else {
+                blks_left->bl_vtotal += kvset_get_nth_vblock_len(ks, j);
+            }
 
             vbcnt++;
-            blks_left->bl_vtotal += kvset_get_nth_vblock_len(ks, j);
         }
 
         if (vbcnt > 0) {
             vbidx_left += vbcnt;
 
-            err = vgmap_vbidx_set(vgmap_src, end - 1, vgmap_left, vbidx_left - 1, vgidx_left);
+            err = vgmap_vbidx_set(vgmap_src, last - 1, vgmap_left, vbidx_left - 1, vgidx_left);
             if (err)
-                return err;
+                goto out;
 
             vgidx_left++;
         }
 
         vbcnt = 0; /* reset vbcnt for the right kvset */
-        if (overlap) {
+        if (metadata->overlaps) {
             /* Append a clone of the overlapping vblock to the right kvset */
-            const uint64_t src_mbid = kvset_get_nth_vblock_id(ks, src_split);
+            off_t off;
             uint64_t clone_mbid;
+            struct mblock_props props;
+            const uint64_t src_mbid = kvset_get_nth_vblock_id(ks, split);
+
+            off = metadata->accessed ? metadata->offset : 0;
+            off = off < PAGE_SIZE ? 0 : roundup(off - PAGE_SIZE, PAGE_SIZE);
 
-            err = mpool_mblock_clone(ks->ks_mp, src_mbid, 0, 0, &clone_mbid);
+            err = mpool_mblock_clone(ks->ks_mp, src_mbid, off, 0, &clone_mbid);
             if (!err) {
                 err = blk_list_append(&blks_right->vblks, clone_mbid);
                 if (!err)
                     err = blk_list_append(result->ks[RIGHT].blks_commit, clone_mbid);
             }
 
             if (err)
-                return err;
+                goto out;
 
-            perfc_rwc++;
-            if (perfc_ison(pc, PERFC_RA_CNCOMP_RBYTES) || perfc_ison(pc, PERFC_RA_CNCOMP_WBYTES)) {
-                struct mblock_props props;
+            log_debug("Cloned mblock (0x%" PRIx64 ") starting at offset %ld", src_mbid, off);
 
-                err = mpool_mblock_props_get(ks->ks_mp, src_mbid, &props);
-                if (!ev(err))
-                    perfc_rwb += props.mpr_write_len;
-                else
-                    err = 0;
-            }
+            err = mpool_mblock_props_get(ks->ks_mp, clone_mbid, &props);
+            if (ev(err))
+                goto out;
+
+            perfc_rwc++;
+            if (perfc_ison(pc, PERFC_RA_CNCOMP_RBYTES) || perfc_ison(pc, PERFC_RA_CNCOMP_WBYTES))
+                perfc_rwb += props.mpr_write_len - off;
 
             vbcnt++;
-            blks_right->bl_vtotal += kvset_get_nth_vblock_len(ks, src_split);
-            src_split++;
+            blks_right->bl_vused -= props.mpr_write_len - props.mpr_alloc_cap;
```

Comment:
Why should we update bl_vused here?

Reply:
Not really sure why I did that since I wrote it so long, but if vused is how much data is referred to by kblocks, then I see no reason to change this and I will update it :).